numpy/_core/src/multiarray/nditer_api.c

```diff
@@ -2157,19 +2157,27 @@ npyiter_copy_to_buffers(NpyIter *iter, char **prev_dataptrs)
             NPY_IT_DBG_PRINT("    unbuffered op (skipping)\n");
             continue;
         }
-        else {
-            /* Pointers shouldn't change (but user could modify them). */
-            user_ptrs[iop] = buffers[iop];
-            NBF_REDUCE_OUTERPTRS(bufferdata)[iop] = buffers[iop];
+
+        /* NOTE: we re-use the NIT_USERPTRS for this check (in iternext) */
+        assert(prev_dataptrs == NULL || prev_dataptrs == NIT_USERPTRS(iter));
+        /*
+         * The prev_dataptrs must match to re-use buffers.  Also coreoffset
+         * must not be 0 to be able to re-use it.
+         * (As of writing coreoffset != 0 implies prev_dataptrs == NULL.)
+         */
+        if (prev_dataptrs == NULL || prev_dataptrs[iop] != dataptrs[iop]) {
+            NIT_OPITFLAGS(iter)[iop] &= ~NPY_OP_ITFLAG_BUF_REUSABLE;
         }
```

Comment:
Admittedly, this now unsets the flag on writeable arguments as well, which is just unnecessary :thinking:.
(Something I'll add to the iterator issue: buffer re-use for writeable ops could be interesting.)